File: app/src/main/java/org/metabrainz/mobile/data/sources/api/HeaderInterceptor.java
package org.metabrainz.mobile.data.sources.api;

import android.content.SharedPreferences;

import org.metabrainz.mobile.presentation.LoginSharedPreferences;

import java.io.IOException;

import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.Response;

public class HeaderInterceptor implements Interceptor {

    private SharedPreferences preferences;

    public HeaderInterceptor(SharedPreferences preferences) {
        this.preferences = preferences;
    }

    @Override
    public Response intercept(Chain chain) throws IOException {
        Request original = chain.request();
        Request request;

        if (original.url().encodedPath().contains("oauth"))
            return chain.proceed(original);
        else {
            if (checkLoginStatus())
                request = original.newBuilder()
                        .header("User-agent", "MusicBrainzAndroid/Test (kartikohri13@gmail.com)")
                        .addHeader("Accept", "application/json")
                        .header("Authorization", " Bearer " + getAccessToken()).build();
            else
                request = original.newBuilder()
                        .header("User-agent", "MusicBrainzAndroid/Test (kartikohri13@gmail.com)")
                        .addHeader("Accept", "application/json").build();

            return chain.proceed(request);
        }
    }

    private boolean checkLoginStatus() {
        return LoginSharedPreferences.getLoginStatus(preferences)
                == LoginSharedPreferences.STATUS_LOGGED_IN;
    }

    private String getAccessToken() {
        return LoginSharedPreferences.getAccessToken(preferences);
    }

}

File: app/src/main/java/org/metabrainz/mobile/data/sources/api/HeaderInterceptor.kt
package org.metabrainz.mobile.data.sources.api

import okhttp3.Interceptor
import okhttp3.Request
import okhttp3.Response
import org.metabrainz.mobile.presentation.features.login.LoginSharedPreferences
import java.io.IOException

internal class HeaderInterceptor : Interceptor {

    @Throws(IOException::class)
    override fun intercept(chain: Interceptor.Chain): Response {
        val original: Request = chain.request()

        // Do not add any headers if fetching data from wiki
        if (original.url.host.contains("wiki")) return chain.proceed(original)

        // Do not add Authorization Header if request is sent to OAuth endpoint except to fetch userinfo
        val request: Request = if (original.url.encodedPath.contains("oauth2")) {
            if (original.url.encodedPath.contains("userinfo")) original.newBuilder()
                    .header("User-agent", "MusicBrainzAndroid/Test (kartikohri13@gmail.com)")
                    .addHeader("Accept", "application/json")
                    .header("Authorization", " Bearer $accessToken").build() else return chain.proceed(original)
        } else {
            if (checkLoginStatus()) {
                original.newBuilder()
                        .header("User-agent", "MusicBrainzAndroid/Test (kartikohri13@gmail.com)")
                        .addHeader("Accept", "application/json")
                        .header("Authorization", " Bearer $accessToken").build()
            } else {
                original.newBuilder()
                        .header("User-agent", "MusicBrainzAndroid/Test (kartikohri13@gmail.com)")
                        .addHeader("Accept", "application/json").build()
            }
        }
        return chain.proceed(request)
    }

    private fun checkLoginStatus(): Boolean {
        return (LoginSharedPreferences.loginStatus == LoginSharedPreferences.STATUS_LOGGED_IN)
    }

    private val accessToken: String get() = LoginSharedPreferences.accessToken!!
}